pkg/istio-agent/xds-proxy.go

```diff
@@ -0,0 +1,329 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istioagent
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"golang.org/x/oauth2"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"google.golang.org/grpc/keepalive"
+	"google.golang.org/grpc/reflection"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	defaultClientMaxReceiveMessageSize = 1024 * 1024 * 8
+	defaultInitialConnWindowSize       = 1024 * 1024 // default gRPC InitialWindowSize
+	defaultInitialWindowSize           = 1024 * 1024 // default gRPC ConnWindowSize
+)
+
+type XdsProxy struct {
+	stopChan             chan struct{}
+	downstreamListener   net.Listener
+	downstreamGrpcServer *grpc.Server
+	upstreamAddress      string
+	upstreamDialOptions  []grpc.DialOption
+}
+
+// XDS Proxy proxies all XDS requests from envoy to istiod, in addition to allowing subsystems inside
+// the agent to also communicate with either istiod/envoy (eg dns, sds, etc).
+// The goal here is to consolidate all xds related connections to istiod/envoy into a
+// single tcp connection with multiple gRPC streams.
+// TODO: Right now, the workloadSDS server and gatewaySDS servers are still separate
+// connections. These need to be consolidated
+func initXdsProxy(sa *Agent) (*XdsProxy, error) {
+	var err error
+	proxy := &XdsProxy{
+		upstreamAddress: sa.proxyConfig.DiscoveryAddress,
+	}
+
+	if err = proxy.initDownstreamServer(); err != nil {
+		return nil, err
+	}
+
+	if proxy.upstreamDialOptions, err = buildUpstreamClientDialOpts(sa); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		_ = proxy.downstreamGrpcServer.Serve(proxy.downstreamListener)
+	}()
+
+	return proxy, nil
+}
+
+// Every time envoy makes a fresh connection to the agent, we reestablish a new connection to the upstream xds
+// This ensures that a new connection between istiod and agent doesn't end up consuming pending messages from envoy
+// as the new connection may not go to the same istiod. Vice versa case also applies.
+func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
+	errChan := make(chan error)
+	requestsChan := make(chan *discovery.DiscoveryRequest)
+	responsesChan := make(chan *discovery.DiscoveryResponse)
+
+	log.Infof("connecting to upstream %s", p.upstreamAddress)
+	upstreamConn, err := grpc.Dial(p.upstreamAddress, p.upstreamDialOptions...)
+	if err != nil {
+		log.Errorf("failed to connect to upstream %s: %v", p.upstreamAddress, err)
+		return err
+	}
+
+	xds := discovery.NewAggregatedDiscoveryServiceClient(upstreamConn)
+	upstream, err := xds.StreamAggregatedResources(context.Background())
```

Comment:
We have no timeout here and the grpc.Dial is async. Does that mean we will hang here forever if there is no connectivity? Will we get adequate logging?

Reply:
good point.. let me investigate